<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#CAIL2021" data-toc-modified-id="CAIL2021-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>CAIL2021</a></span></li><li><span><a href="#题库情况" data-toc-modified-id="题库情况-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>题库情况</a></span></li><li><span><a href="#关于题库token2id与参考书目文档token2id对比" data-toc-modified-id="关于题库token2id与参考书目文档token2id对比-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>关于题库token2id与参考书目文档token2id对比</a></span><ul class="toc-item"><li><span><a href="#读取TOKEN2ID，TOKEN2FREQUENCY与REFERENCE_TOKEN2ID，REFERENCE_TOKEN2FREQUENCY并转为字典" data-toc-modified-id="读取TOKEN2ID，TOKEN2FREQUENCY与REFERENCE_TOKEN2ID，REFERENCE_TOKEN2FREQUENCY并转为字典-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>读取TOKEN2ID，TOKEN2FREQUENCY与REFERENCE_TOKEN2ID，REFERENCE_TOKEN2FREQUENCY并转为字典</a></span></li><li><span><a href="#统计那些在参考书目中但不在题库的分词" data-toc-modified-id="统计那些在参考书目中但不在题库的分词-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>统计那些在参考书目中但不在题库的分词</a></span></li><li><span><a href="#统计那些在题库中但不在参考书目的分词" data-toc-modified-id="统计那些在题库中但不在参考书目的分词-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>统计那些在题库中但不在参考书目的分词</a></span></li><li><span><a href="#词频统计信息" data-toc-modified-id="词频统计信息-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>词频统计信息</a></span></li></ul></li><li><span><a href="#gensim模型" data-toc-modified-id="gensim模型-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>gensim模型</a></span><ul class="toc-item"><li><span><a href="#tfidf模型不同参数的情况" data-toc-modified-id="tfidf模型不同参数的情况-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>tfidf模型不同参数的情况</a></span></li><li><span><a href="#Embedding模型测试" data-toc-modified-id="Embedding模型测试-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Embedding模型测试</a></span></li></ul></li><li><span><a href="#BERT模型测试" data-toc-modified-id="BERT模型测试-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>BERT模型测试</a></span></li><li><span><a href="#参考书目文档的高阶处理" data-toc-modified-id="参考书目文档的高阶处理-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>参考书目文档的高阶处理</a></span><ul class="toc-item"><li><span><a href="#统计字符种类及频率" data-toc-modified-id="统计字符种类及频率-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>统计字符种类及频率</a></span></li><li><span><a href="#存储解析树" data-toc-modified-id="存储解析树-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>存储解析树</a></span></li><li><span><a href="#句法树数据挖掘" data-toc-modified-id="句法树数据挖掘-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>句法树数据挖掘</a></span><ul class="toc-item"><li><span><a href="#关于句法树标注" data-toc-modified-id="关于句法树标注-6.3.1"><span class="toc-item-num">6.3.1&nbsp;&nbsp;</span>关于句法树标注</a></span></li><li><span><a href="#句法树标注情况分析（以reference_parse_tree为例）" data-toc-modified-id="句法树标注情况分析（以reference_parse_tree为例）-6.3.2"><span class="toc-item-num">6.3.2&nbsp;&nbsp;</span>句法树标注情况分析（以reference_parse_tree为例）</a></span></li><li><span><a href="#句法树标注情况分析（以question_parse_tree为例）" data-toc-modified-id="句法树标注情况分析（以question_parse_tree为例）-6.3.3"><span class="toc-item-num">6.3.3&nbsp;&nbsp;</span>句法树标注情况分析（以question_parse_tree为例）</a></span></li><li><span><a href="#句法树的基本情况的一些特点挖掘" data-toc-modified-id="句法树的基本情况的一些特点挖掘-6.3.4"><span class="toc-item-num">6.3.4&nbsp;&nbsp;</span>句法树的基本情况的一些特点挖掘</a></span></li><li><span><a href="#句法树生成networkx图测试" data-toc-modified-id="句法树生成networkx图测试-6.3.5"><span class="toc-item-num">6.3.5&nbsp;&nbsp;</span>句法树生成networkx图测试</a></span></li><li><span><a href="#句法树分支数量情况（最大子女数量）" data-toc-modified-id="句法树分支数量情况（最大子女数量）-6.3.6"><span class="toc-item-num">6.3.6&nbsp;&nbsp;</span>句法树分支数量情况（最大子女数量）</a></span></li><li><span><a href="#句法树在Dataset中的生成情况测试" data-toc-modified-id="句法树在Dataset中的生成情况测试-6.3.7"><span class="toc-item-num">6.3.7&nbsp;&nbsp;</span>句法树在Dataset中的生成情况测试</a></span></li><li><span><a href="#use_reference为False的测试" data-toc-modified-id="use_reference为False的测试-6.3.8"><span class="toc-item-num">6.3.8&nbsp;&nbsp;</span>use_reference为False的测试</a></span></li><li><span><a href="#use_reference为True的测试" data-toc-modified-id="use_reference为True的测试-6.3.9"><span class="toc-item-num">6.3.9&nbsp;&nbsp;</span>use_reference为True的测试</a></span></li></ul></li><li><span><a href="#依存关系图" data-toc-modified-id="依存关系图-6.4"><span class="toc-item-num">6.4&nbsp;&nbsp;</span>依存关系图</a></span></li></ul></li><li><span><a href="#BERT输出本地化的一些测试" data-toc-modified-id="BERT输出本地化的一些测试-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>BERT输出本地化的一些测试</a></span></li><li><span><a href="#图模型测试" data-toc-modified-id="图模型测试-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>图模型测试</a></span></li></ul></div>

# CAIL2021

In [1]:
import os
import re
import json
import time
import dill
import torch
import numpy
import gensim
import pandas


from gensim.corpora import Dictionary, MmCorpus
from matplotlib import pyplot as plt
from collections import Counter

from gensim.models import Word2Vec, FastText, Doc2Vec, WordEmbeddingSimilarityIndex
from gensim.models.doc2vec import TaggedDocument
from gensim.similarities import SoftCosineSimilarity, SparseTermSimilarityMatrix



from config import DatasetConfig, RetrievalModelConfig, EmbeddingModelConfig
from setting import *
from preprocess import build_gensim_retrieval_models

from src.dataset import BasicDataset, generate_dataloader
from src.retrieval_model import GensimRetrievalModel
from src.embedding_model import GensimEmbeddingModel
from src.evaluation_tools import evaluate_gensim_model_in_filling_subject
from src.graph import Graph
from src.graph_tools import generate_pos_tags, generate_parse_tree, generate_pos_tags_from_parse_tree, parse_tree_to_graph
from src.qa_module import TreeNodeEncoder
from src.utils import load_args, timer, is_chinese, is_number, is_alphabet, is_symbol

%matplotlib inline

Using backend: pytorch


ImportError: cannot import name 'generate_dataloader' from 'src.dataset' (f:\code\project\cail2021\src\dataset.py)

# 题库情况

In [ ]:
for path in RAW_TRAINSET_PATHs:
    data = []
    with open(path, 'r', encoding='utf8') as f:
        while True:
            line = f.readline()
            if not line:
                break
            data.append(json.loads(line))
    print(len(data))
    
for path in RAW_TESTSET_PATHs:
    data = []
    with open(path, 'r', encoding='utf8') as f:
        while True:
            line = f.readline()
            if not line:
                break
            data.append(json.loads(line))
    print(len(data))



In [ ]:
print(7775 + 13297)
print(1986 + 3303)

# 关于题库token2id与参考书目文档token2id对比

## 读取TOKEN2ID，TOKEN2FREQUENCY与REFERENCE_TOKEN2ID，REFERENCE_TOKEN2FREQUENCY并转为字典

In [ ]:
token2id_df = pandas.read_csv(TOKEN2ID_PATH, sep='\t', header=0)
token2frequency_df = pandas.read_csv(TOKEN2FREQUENCY_PATH, sep='\t', header=0)
reference_token2id_df = pandas.read_csv(REFERENCE_TOKEN2ID_PATH, sep='\t', header=0)
reference_token2frequency_df = pandas.read_csv(REFERENCE_TOKEN2FREQUENCY_PATH, sep='\t', header=0)

In [ ]:
token2id = {token: _id for token, _id in zip(token2id_df['token'], token2id_df['id'])}
token2frequency = {token: frequency for token, frequency in zip(token2frequency_df['token'], token2frequency_df['frequency'])}
reference_token2id = {token: _id for token, _id in zip(reference_token2id_df['token'], reference_token2id_df['id'])}
reference_token2frequency = {token: frequency for token, frequency in zip(reference_token2frequency_df['token'], reference_token2frequency_df['frequency'])}

## 统计那些在参考书目中但不在题库的分词

结论：前三个都是标点符号，后面的词频都非常低

In [ ]:
count = 0
token2frequency_extra_in_reference = {}
for token in reference_token2id:
    if not token in token2id:
        count += 1
        token2frequency_extra_in_reference[token] = reference_token2frequency[token]

print(len(token2frequency_extra_in_reference))
print(Counter(token2frequency_extra_in_reference).most_common(100))

## 统计那些在题库中但不在参考书目的分词

结论：最多的是顿号，然后大都是名词性短语（如人名，地名等）

In [ ]:
count = 0
token2frequency_not_in_reference = {}
for token in token2id:
    if not token in reference_token2id:
        count += 1
        token2frequency_not_in_reference[token] = token2frequency[token]

print(len(token2frequency_not_in_reference))
print(Counter(token2frequency_not_in_reference).most_common(100))

## 词频统计信息

In [ ]:
counter = Counter(token2frequency)
reference_Counter = Counter(reference_token2frequency)

def _plot_frequency(_counter):
    _y = numpy.array(list(map(lambda x: x[1], counter.most_common())), dtype=numpy.int64)
    _x = numpy.array(list(range(_y.shape[0])))
    
    
    for i in range(10):
        plt.plot(_x[i * 100: (i + 1) * 100], _y[i * 100: (i + 1) * 100])
        plt.show()

_plot_frequency(counter)
_plot_frequency(reference_Counter)

In [ ]:
token2frequency_df['frequency'].describe()

In [ ]:
reference_token2frequency_df['frequency'].describe()

# gensim模型

## tfidf模型不同参数的情况

目前看下来tfidf的效果最好，logentropy也不错，但是它没什么参数可调，word2vec和fasttext效果较好，但不如tfidf，并且我把词向量维数从256调为512后还是没有显著的性能提升，hdp效果一般，lsi较差，lda非常差，因此我只考虑在tfidf上做测试了，主要是smartirs, slope, pivot三个参数的调整。

目前看下来hit@3的水平最高还是只有91.7%(6438个正确)，跟论文中比hit@1的水平基本持平TextCNN，但是hit@3要差BERT一点点，但是要经济得多

tfidf调参结果是6438个，logentropy为6381个正确预测，总共7021个

In [ ]:
summarys = json.load(open(os.path.join(TEMP_DIR, 'test_smartirs.json'), 'r')) 

In [ ]:
summary_dict = {
    'smartirs': [],
    'pivot': [],
    'slope': [],
    'hit@1': [],
    'hit@3': [],
    'hit@5': [],
    'hit@10': [],
}
for summary in summarys:
    
    summary_dict['smartirs'].append(summary['args']['smartirs_tfidf'])
    summary_dict['pivot'].append(summary['args']['pivot'])
    summary_dict['slope'].append(summary['args']['slope'])
    summary_dict['hit@1'].append(summary['result']['tfidf']['hit@1'])
    summary_dict['hit@3'].append(summary['result']['tfidf']['hit@3'])
    summary_dict['hit@5'].append(summary['result']['tfidf']['hit@5'])
    summary_dict['hit@10'].append(summary['result']['tfidf']['hit@10'])
    
summary_dataframe = pandas.DataFrame(summary_dict, columns=list(summary_dict.keys()))

In [ ]:
print(summary_dataframe['hit@1'].max() / 7021)
print(summary_dataframe['hit@3'].max() / 7021)
print(summary_dataframe['hit@5'].max() / 7021)
print(summary_dataframe['hit@10'].max() / 7021)

In [ ]:
print(summary_dataframe['hit@3'].max())
print(summary_dataframe[summary_dataframe['hit@3'] == 6438])

In [ ]:
summary_dataframe.to_csv('summary.csv', header=True, index=False, sep='\t')

## Embedding模型测试

In [ ]:
model_word2vec = gensim.models.Word2Vec.load(REFERENCE_WORD2VEC_MODEL_PATH)
model_fasttext = gensim.models.Word2Vec.load(REFERENCE_FASTTEXT_MODEL_PATH)

print(model_word2vec.wv.similarity('抢劫', '强奸'))
print(model_fasttext.wv.similarity('抢劫', '强奸'))
print(model_word2vec.wv.similarity('绑架', '盗窃'))
print(model_fasttext.wv.similarity('绑架', '盗窃'))

In [ ]:
model_word2vec.wv.index2word

In [ ]:
similarity_index = WordEmbeddingSimilarityIndex(model.wv)

In [ ]:
dictionary = Dictionary.load(REFERENCE_DICTIONARY_PATH)
similarity_matrix = SparseTermSimilarityMatrix(similarity_index, dictionary)

In [ ]:
help(similarity_matrix)

In [ ]:
similarity_matrix[0, 0]

In [ ]:
model.wv.similarity('抢劫', '司法')

In [ ]:
len(model.wv.index2word)

In [ ]:
model.wv.similar_by_word

In [ ]:
model.wv['强奸', '盗窃'].shape

In [ ]:
model.wv.__getitem__('人民法院').shape

In [ ]:
'人民法院aaa' in model.wv

In [ ]:
type(model.wv.__getitem__('人民法院'))

In [ ]:
model = Doc2Vec.load(REFERENCE_DOC2VEC_MODEL_PATH)

In [ ]:
model.infer_vector(['在', '法定代表', '人', '和', '法人', '关系', '的', '问题', '上', '，', '下列', '哪些', '表述', '是', '正确', '的']).shape

# BERT模型测试

In [ ]:
# -*- coding: utf-8 -*-
# @author: caoyang
# @email: caoyang@163.sufe.edu.cn

import os
import json
import time
import pandas

from config import DatasetConfig, RetrievalModelConfig, EmbeddingModelConfig
from setting import *

from src.dataset import Dataset
from src.retrieval_model import GensimRetrievalModel
from src.embedding_model import GensimEmbeddingModel, TransformersEmbeddingModel
from src.evaluation_tools import evaluate_gensim_model_in_filling_subject
from src.plot_tools import train_plot_choice, train_plot_judgment
from src.utils import load_args

args = load_args(Config=EmbeddingModelConfig)

# tem = TransformersEmbeddingModel(args=args)
# tokenizer, model = tem.load_bert_model()

tokenizer, model = TransformersEmbeddingModel.load_bert_model()



text = [
	'劳动法有广义和狭义之分:狭义上的劳动法，一般是指国家最高立法机构制定颁布的全国性.练合性的劳动法，即《中华人民共和国劳动法》(以下简称《劳动法》)；广义上的劳动法，是指调整劳动关系以及与劳动关系有密切联系的其他社会关系的法律规范的总称。劳动法有广义和狭义之分:狭义上的劳动法，一般是指国家最高立法机构制定颁布的全国性.练合性的劳动法，即《中华人民共和国劳动法》(以下简称《劳动法》)；广义上的劳动法，是指调整劳动关系以及与劳动关系有密切联系的其他社会关系的法律规范的总称。劳动法有广义和狭义之分:狭义上的劳动法，一般是指国家最高立法机构制定颁布的全国性.练合性的劳动法，即《中华人民共和国劳动法》(以下简称《劳动法》)；广义上的劳动法，是指调整劳动关系以及与劳动关系有密切联系的其他社会关系的法律规范的总称。劳动法有广义和狭义之分:狭义上的劳动法，一般是指国家最高立法机构制定颁布的全国性.练合性的劳动法，即《中华人民共和国劳动法》(以下简称《劳动法》)；广义上的劳动法，是指调整劳动关系以及与劳动关系有密切联系的其他社会关系的法律规范的总称。劳动法有广义和狭义之分:狭义上的劳动法，一般是指国家最高立法机构制定颁布的全国性.练合性的劳动法，即《中华人民共和国劳动法》(以下简称《劳动法》)；广义上的劳动法，是指调整劳动关系以及与劳动关系有密切联系的其他社会关系的法律规范的总称。劳动法有广义和狭义之分:狭义上的劳动法，一般是指国家最高立法机构制定颁布的全国性.练合性的劳动法，即《中华人民共和国劳动法》(以下简称《劳动法》)；广义上的劳动法，是指调整劳动关系以及与劳动关系有密切联系的其他社会关系的法律规范的总称。劳动法有广义和狭义之分:狭义上的劳动法，一般是指国家最高立法机构制定颁布的全国性.练合性的劳动法，即《中华人民共和国劳动法》(以下简称《劳动法》)；广义上的劳动法，是指调整劳动关系以及与劳动关系有密切联系的其他社会关系的法律规范的总称。劳动法有广义和狭义之分:狭义上的劳动法，一般是指国家最高立法机构制定颁布的全国性.练合性的劳动法，即《中华人民共和国劳动法》(以下简称《劳动法》)；广义上的劳动法，是指调整劳动关系以及与劳动关系有密切联系的其他社会关系的法律规范的总称。劳动法有广义和狭义之分:狭义上的劳动法，一般是指国家最高立法机构制定颁布的全国性.练合性的劳动法，即《中华人民共和国劳动法》(以下简称《劳动法》)；广义上的劳动法，是指调整劳动关系以及与劳动关系有密切联系的其他社会关系的法律规范的总称。劳动法有广义和狭义之分:狭义上的劳动法，一般是指国家最高立法机构制定颁布的全国性.练合性的劳动法，即《中华人民共和国劳动法》(以下简称《劳动法》)；广义上的劳动法，是指调整劳动关系以及与劳动关系有密切联系的其他社会关系的法律规范的总称。劳动法有广义和狭义之分:狭义上的劳动法，一般是指国家最高立法机构制定颁布的全国性.练合性的劳动法，即《中华人民共和国劳动法》(以下简称《劳动法》)；广义上的劳动法，是指调整劳动关系以及与劳动关系有密切联系的其他社会关系的法律规范的总称。劳动法有广义和狭义之分:狭义上的劳动法，一般是指国家最高立法机构制定颁布的全国性.练合性的劳动法，即《中华人民共和国劳动法》(以下简称《劳动法》)；广义上的劳动法，是指调整劳动关系以及与劳动关系有密切联系的其他社会关系的法律规范的总称。劳动法有广义和狭义之分:狭义上的劳动法，一般是指国家最高立法机构制定颁布的全国性.练合性的劳动法，即《中华人民共和国劳动法》(以下简称《劳动法》)；广义上的劳动法，是指调整劳动关系以及与劳动关系有密切联系的其他社会关系的法律规范的总称。劳动法有广义和狭义之分:狭义上的劳动法，一般是指国家最高立法机构制定颁布的全国性.练合性的劳动法，即《中华人民共和国劳动法》(以下简称《劳动法》)；广义上的劳动法，是指调整劳动关系以及与劳动关系有密切联系的其他社会关系的法律规范的总称。劳动法有广义和狭义之分:狭义上的劳动法，一般是指国家最高立法机构制定颁布的全国性.练合性的劳动法，即《中华人民共和国劳动法》(以下简称《劳动法》)；广义上的劳动法，是指调整劳动关系以及与劳动关系有密切联系的其他社会关系的法律规范的总称。',
] * 1

encoded_input = tokenizer(text, return_tensors='pt', padding=True, truncation=True, max_length=512)


print(encoded_input)

t = time.time()
for i in range(1):
	output = model(**encoded_input)
print(time.time() - t)

input('pause')


print('#' * 64)
print('last_hidden_state: ', output.get('last_hidden_state').shape)
print('pooler_output', output.get('pooler_output').shape)
print('cross_attentions', output.cross_attentions)
print('hidden_states', output.hidden_states)
print('#' * 64)

input()

for encoded_hidden_state in output.last_hidden_state:
	i = 0
	for layer in encoded_hidden_state:
		i += 1
		print(i, layer)
		input()
		

print('#' * 64)
for encoded_text in output.pooler_output:
	print(encoded_text)
	print('-*' * 32)

'''
last_hidden_state:  torch.Size([1, 120, 768])
pooler_output torch.Size([1, 768])
'''

# print(output.logits)

['__annotations__', '__class__', '__contains__', '__dataclass_fields__', 
'__dataclass_params__', '__delattr__', '__delitem__', '__dict__', 
'__dir__', '__doc__', '__eq__', '__format__', '__ge__', 
'__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', 
'__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', 
'__module__', '__ne__', '__new__', '__post_init__', '__reduce__', 
'__reduce_ex__', '__repr__', '__reversed__', '__setattr__', '__setitem__', 
'__sizeof__', '__str__', '__subclasshook__', 
'attentions', 'clear', 'copy', 'cross_attentions', 'fromkeys', 'get', 
'hidden_states', 'items', 'keys', 'last_hidden_state', 'move_to_end', 
'past_key_values', 'pooler_output', 'pop', 'popitem', 'setdefault', 
'to_tuple', 'update', 'values']


# tokenizer = AutoTokenizer.from_pretrained("bert-base-cased-finetuned-mrpc")
# model = AutoModelForSequenceClassification.from_pretrained("bert-base-cased-finetuned-mrpc")
# classes = ["not paraphrase", "is paraphrase"]
# sequence_0 = "The company HuggingFace is based in New York City"
# sequence_1 = "Apples are especially bad for your health"
# sequence_2 = "HuggingFace's headquarters are situated in Manhattan"
# paraphrase = tokenizer(sequence_0, sequence_2, return_tensors="pt")
# not_paraphrase = tokenizer(sequence_0, sequence_1, return_tensors="pt")
# paraphrase_classification_logits = model(**paraphrase).logits
# not_paraphrase_classification_logits = model(**not_paraphrase).logits
# paraphrase_results = torch.softmax(paraphrase_classification_logits, dim=1).tolist()[0]
# not_paraphrase_results = torch.softmax(not_paraphrase_classification_logits, dim=1).tolist()[0]
# # Should be paraphrase
# for i in range(len(classes)):
	# print(f"{classes[i]}: {int(round(paraphrase_results[i] * 100))}%")
# # Should not be paraphrase
# for i in range(len(classes)):
	# print(f"{classes[i]}: {int(round(not_paraphrase_results[i] * 100))}%")



# 参考书目文档的高阶处理

## 统计字符种类及频率

In [ ]:
reference_dataframe = pandas.read_csv(REFERENCE_PATH, sep='\t', header=0)

numbers = []
alphabets = []
symbols = []

for i in range(reference_dataframe.shape[0]):
    content_tokens = eval(reference_dataframe.loc[i, 'content'])
    for token in content_tokens:
        if is_chinese(token):
            continue
        elif is_number(token):
            # print(f'number: {token}')
            numbers.append(token)
        elif is_alphabet(token):
            # print(f'alphabet: {token}')
            alphabets.append(token)
        else:
            # print(f'symbol: {token}')
            symbols.append(token)
            
# print(set(numbers))
# print(set(alphabets))
print(set(symbols))
# print(Counter(numbers).most_common())
# print(Counter(alphabets).most_common())
print(Counter(symbols).most_common())

In [ ]:
print(reference_dataframe.shape)

## 存储解析树

In [ ]:
tree_dataframe = pandas.read_csv(REFERENCE_PARSE_TREE_PATH, sep='\t', header=0, dtype=str)
reference_dataframe = pandas.read_csv(REFERENCE_PATH, sep='\t', header=0, dtype=str)
reference_dataframe = reference_dataframe.fillna('')

In [ ]:
print(tree_dataframe.shape)
print(reference_dataframe.shape)

In [ ]:
tree_dataframe.head(2)

In [ ]:
tree_dataframe['error'] = tree_dataframe['parse_tree'].map(lambda x: not x[:7] == "['(ROOT")

In [ ]:
tree_dataframe[tree_dataframe['error']].shape

In [ ]:
tree_dataframe[tree_dataframe['error']]

In [ ]:
content_tokens = eval(reference_dataframe.loc[19592, 'content'])

In [ ]:
content_tokens

## 句法树数据挖掘

- reference句法树的词性标注（即叶子节点上的标注）共计33种，具体解释见[https://blog.csdn.net/weixin_30642561/article/details/97772970?spm=1001.2101.3001.6650.10&utm_medium=distribute.pc_relevant.none-task-blog-2%7Edefault%7ECTRLIST%7ERate-10.pc_relevant_antiscanv2&depth_1-utm_source=distribute.pc_relevant.none-task-blog-2%7Edefault%7ECTRLIST%7ERate-10.pc_relevant_antiscanv2&utm_relevant_index=14](https://blog.csdn.net/weixin_30642561/article/details/97772970?spm=1001.2101.3001.6650.10&utm_medium=distribute.pc_relevant.none-task-blog-2%7Edefault%7ECTRLIST%7ERate-10.pc_relevant_antiscanv2&depth_1-utm_source=distribute.pc_relevant.none-task-blog-2%7Edefault%7ECTRLIST%7ERate-10.pc_relevant_antiscanv2&utm_relevant_index=14)

- reference句法树的词性标注（即叶子节点上的标注）共计28种，具体解释见[https://blog.csdn.net/zkq_1986/article/details/93175100](https://blog.csdn.net/zkq_1986/article/details/93175100)（不确定）

- jieba的标注可见[https://blog.csdn.net/kevin_darkelf/article/details/39520881](https://blog.csdn.net/kevin_darkelf/article/details/39520881)

### 关于句法树标注

In [ ]:
s = """ROOT：要处理文本的语句

IP：简单从句

NP：名词短语

VP：动词短语

PU：断句符，通常是句号、问号、感叹号等标点符号

LCP：方位词短语

PP：介词短语

CP：由‘的’构成的表示修饰性关系的短语

DNP：由‘的’构成的表示所属关系的短语

ADVP：副词短语

ADJP：形容词短语

DP：限定词短语

QP：量词短语

NN：常用名词

NR：固有名词

NT：时间名词

PN：代词

VV：动词

VC：是

CC：表示连词

VE：有

VA：表语形容词

AS：内容标记（如：了）

VRD：动补复合词

CD: 表示基数词

DT: determiner 表示限定词

EX: existential there 存在句

FW: foreign word 外来词

IN: preposition or conjunction, subordinating 介词或从属连词

JJ: adjective or numeral, ordinal 形容词或序数词

JJR: adjective, comparative 形容词比较级

JJS: adjective, superlative 形容词最高级

LS: list item marker 列表标识

MD: modal auxiliary 情态助动词

PDT: pre-determiner 前位限定词

POS: genitive marker 所有格标记

PRP: pronoun, personal 人称代词

RB: adverb 副词

RBR: adverb, comparative 副词比较级

RBS: adverb, superlative 副词最高级

RP: particle 小品词

SYM: symbol 符号

TO:”to” as preposition or infinitive marker 作为介词或不定式标记

WDT: WH-determiner WH限定词

WP: WH-pronoun WH代词

WP$: WH-pronoun, possessive WH所有格代词

WRB:Wh-adverb WH副词"""

lines = list(filter(None, s.splitlines()))
labels = [line.replace('：', ':').split(':')[0].replace(' ', '') for line in lines]
print(len(labels))
print(len(set(labels)))
labels = set(labels)

mylabels = set(['ADJP', 'ADVP', 'CLP', 'CP', 'DFL', 'DNP', 'DP', 'DVP', 'FLR', 'FRAG', 'INC', 'INTJ', 'IP', 'LCP', 'LST', 'NP', 'PP', 'PRN', 'QP', 'ROOT', 'UCP', 'VCD', 'VCP', 'VNV', 'VP', 'VPT', 'VRD', 'VSB'])

print(mylabels - labels)

### 句法树标注情况分析（以reference_parse_tree为例）

In [ ]:
df = pandas.read_csv(REFERENCE_PARSE_TREE_PATH, sep='\t', header=0)
df['id'] = df['id'].astype(int)

token2id = pandas.read_csv(REFERENCE_TOKEN2ID_PATH, sep='\t', header=0)

In [ ]:
token2id

In [ ]:
token2id[token2id['token'] == '\xa0']

In [ ]:
# 寻找解析错误的语句: 解析正确的语句开头一定是['(ROOT
for i, content in enumerate(df['content']):
    if not content.startswith("['(ROOT"):
        print(i, content)

In [ ]:
# 统计所有的句法树标注
labels1 = []

for i, content in enumerate(df['content']):
    parse_trees = eval(content)
    for parse_tree in parse_trees:
        string = parse_tree.replace('(', ' ').replace(')', ' ').strip()
        tokens = string.split()
        labels1.extend(tokens)


In [ ]:
# 统计所有的句法树标注（排除所有叶子节点内容，我感觉叶子节点标注的是词性，而非句法树标注，使用正则替换消除）

labels2 = []
regex_compiler = re.compile(r'\([^\(\)]+\)', re.I)
for i, content in enumerate(df['content']):
    parse_trees = eval(content)
    for parse_tree in parse_trees:
        string = regex_compiler.sub('', parse_tree)
        string = string.replace('(', ' ').replace(')', ' ').strip()
        tokens = string.split()
        labels2.extend(tokens)


In [ ]:
counter = Counter(labels2)

In [ ]:
counter.most_common()

In [ ]:
error = ['INTJ', 'INC', 'DFL', 'LST', 'VCP', 'PRN', 'VPT', 'VNV', 'UCP', 'VCD', 'VSB', 'FRAG', 'DVP', 'FLR', 'CLP']
for x in error:
    print(x, counter.get(x))

In [ ]:
# 统计所有的句法树标注（只抓叶子节点中的内容，我想看看跟上面一个标注是否有重复）
labels3 = []
regex_compiler = re.compile(r'\([^\(\)]+\)', re.I)
for i, content in enumerate(df['content']):
    parse_trees = eval(content)
    for parse_tree in parse_trees:
        results = regex_compiler.findall(parse_tree)
        for result in results:
            assert result[0] == '('
            labels3.append(result.split()[0][1:])

In [ ]:
labels1 = set(labels1)
labels2 = set(labels2)
labels3 = set(labels3)
reference_tokens = token2id['token'].unique()
reference_tokens = set(reference_tokens)
print(len(labels1))
print(len(labels2))
print(len(labels3))
print(len(reference_tokens))

In [ ]:
true_labels1 = labels1 - reference_tokens
true_labels2 = labels2 - reference_tokens
true_labels3 = labels3 - reference_tokens
print(len(true_labels1))
print(len(true_labels2))
print(len(true_labels3))

In [ ]:
print(labels2.intersection(reference_tokens))
print(labels3.intersection(reference_tokens))
print(labels2.intersection(labels3)) # 是空集，这样的结果非常赞

In [ ]:
print(sorted(labels2))
print(sorted(labels3))

In [ ]:
all_labels = labels2.union(labels3)
print(len(all_labels))

In [ ]:
LABELS2.issubset(all_labels)

In [ ]:
all_labels - LABELS2

In [ ]:
LABELS2 = {'ROOT', 'IP', 'NP', 'VP', 'PU', 'LCP', 'PP', 'CP', 'DNP', 'ADVP', 'ADJP', 'DP', 'QP', 'NN', 'NR', 'NT', 'PN', 'VV', 'VC', 'VA', 'CC', 'VE', 'AS', 'VRD'}

In [ ]:
LABELS2 - labels2

In [ ]:
LABELS2 - labels3

In [ ]:
labels2 - LABELS2

In [ ]:
# 出现在reference_tokens中但没有出现在句法解析树中的token（这是不正常的，因为理论上reference_tokens都出现在解析树中了）
reference_tokens - labels

In [ ]:
labels

In [ ]:
true_labels

### 句法树标注情况分析（以question_parse_tree为例）

In [ ]:
df = pandas.read_csv(QUESTION_PARSE_TREE_PATH, sep='\t', header=0, dtype=str)

In [ ]:
df[df['id'] == '1_3906']['statement'].tolist()

QUESTION_PARSE_TREE中的确没有解析错误的情况

In [ ]:
# 寻找解析错误的语句: 解析正确的语句开头一定是['(ROOT
for i in range(df.shape[0]):
    _id = df.loc[i, 'id']
    statement = df.loc[i, 'statement']
    option_a = df.loc[i, 'option_a']
    option_b = df.loc[i, 'option_b']
    option_c = df.loc[i, 'option_c']
    option_d = df.loc[i, 'option_d']
    if not (statement.startswith("['(ROOT") or statement.startswith('["(ROOT')):
        print(_id, 'statement', statement)
    if not (option_a.startswith("['(ROOT") or option_a.startswith('["(ROOT')):
        print(_id, 'option_a', option_a)
    if not (option_b.startswith("['(ROOT") or option_b.startswith('["(ROOT')):
        print(_id, 'option_b', option_b)
    if not (option_c.startswith("['(ROOT") or option_c.startswith('["(ROOT')):
        print(_id, 'option_c', option_c)
    if not (option_d.startswith("['(ROOT") or option_d.startswith('["(ROOT')):
        print(_id, 'option_d', option_d)
        

In [ ]:
# 是否可以不使用eval，直接进行generate_pos_tags_from_parse_tree?

def _f(_x):
    _x = eval(_x)
    _res = []
    for _i in _x:
        _res.extend(generate_pos_tags_from_parse_tree(_i))
    
    return _res

array1 = df['statement'].map(generate_pos_tags_from_parse_tree)
array2 = df['statement'].map(_f)

In [ ]:
# 答：不可以，因为eval之后可以预先处理掉一些转义符
# 下面两个数字不相等的原因就是不eval会把转义符识别为字符

print(len(array1))
print(sum(array1==array2))

In [ ]:
for a, b in zip(array1, array2):
    if not a == b:
        print(a)
        print(b)
        print('#' * 64)

In [ ]:
# 统计所有的句法树标注（排除所有叶子节点内容，我感觉叶子节点标注的是词性，而非句法树标注，使用正则替换消除）
labels2 = []
regex_compiler = re.compile(r'\([^\(\)]+\)', re.I)

for column in ['statement', 'option_a', 'option_b', 'option_c', 'option_d']:
    for i, content in enumerate(df[column]):
        parse_trees = eval(content)
        for parse_tree in parse_trees:
            string = regex_compiler.sub('', parse_tree)
            string = string.replace('(', ' ').replace(')', ' ').strip()
            tokens = string.split()
            labels2.extend(tokens)


In [ ]:
# 统计所有的句法树标注（只抓叶子节点中的内容，我想看看跟上面一个标注是否有重复）
labels3 = []
regex_compiler = re.compile(r'\([^\(\)]+\)', re.I)
for column in ['statement', 'option_a', 'option_b', 'option_c', 'option_d']:
    for i, content in enumerate(df[column]):
        parse_trees = eval(content)
        for parse_tree in parse_trees:
            results = regex_compiler.findall(parse_tree)
            for result in results:
                assert result[0] == '('
                labels3.append(result.split()[0][1:])

In [ ]:
labels2 = sorted(list(set(labels2)))
labels3 = sorted(list(set(labels3)))

print(f'句法树标注({len(labels2)}): {labels2}')
print(f'词性标注({len(labels3)}): {labels3}')

说明QUESTION_PARSE_TREE里的标注与REFERENCE_PARSE_TREE是完全相同的，只是句法树标注少一个INTJ（）

In [ ]:
for i in labels2:
    assert i in STANFORD_SYNTACTIC_TAG, i

for j in labels3:
    assert j in STANFORD_POS_TAG, j
    
for i in STANFORD_SYNTACTIC_TAG:
    assert i in labels2, i

### 句法树的基本情况的一些特点挖掘

In [13]:
# 是否保留原序列的顺序？（是）

reference_df = pandas.read_csv(REFERENCE_PATH, sep='\t')
reference_parse_tree_df = pandas.read_csv(REFERENCE_PARSE_TREE_PATH, sep='\t', header=0)

reference_df['content'] = reference_df['content'].map(eval)

print(reference_df.shape)
print(reference_df.columns)
print(reference_parse_tree_df.shape)
print(reference_parse_tree_df.columns)

assert reference_df.shape[0] == reference_parse_tree_df.shape[0]

(24718, 5)
Index(['law', 'chapter_number', 'chapter_name', 'section', 'content'], dtype='object')
(24718, 2)
Index(['id', 'content'], dtype='object')


1. 下面的脚本验证了句法树中的分词序列位置相较于原序列是没有发生改变的，不过有几个特殊符号的替换：

   - 原分词序列中的'('与')'队以ing句法树中的'-LRB-'与'-RRB-'
   - 原分词序列中的'\xa0'（这个其实就是空格，只有jieba分词会出现这个东西），在句法树中的分词对应为空

2. 下面的脚本说明只要做如上的几个替换，句法树里的分词序列与原序列是完全相同的

In [17]:
error = []

for i in range(reference_df.shape[0]):
    content = reference_df.loc[i, 'content']
    content_from_parse_tree = generate_pos_tags_from_parse_tree(parse_tree=reference_parse_tree_df.loc[i, 'content'])

    content_ex = list(map(lambda x: x[1], content_from_parse_tree))
    content_ex = list(map(lambda x: '(' if x == '-LRB-' else ')' if x == '-RRB-' else '\xa0' if x == '' else x, content_ex))
    # 若使用下面的方法，则会跳出几个error
    # content_ex = list(map(lambda x: '(' if x == '-LRB-' else ')' if x == '-RRB-' else x, content_ex))           

    if content == content_ex:
        continue
    else:
        print(i)
        error.append(i)

print(f'参考文档中不能匹配的数量: {len(error)}')

参考文档中不能匹配的数量: 0


下面：类似地在question数据集上也进行与reference相同地测试，结论是相同的

In [18]:
# 是否保留原序列的顺序？（是）

filepaths = TRAINSET_PATHs + VALIDSET_PATHs + TESTSET_PATHs

question_df = pandas.concat([pandas.read_csv(filepath, sep='\t', header=0) for filepath in filepaths]).reset_index(drop=True)
print(question_df.shape)

question_parse_tree_df = pandas.read_csv(QUESTION_PARSE_TREE_PATH, sep='\t', header=0)
print(question_parse_tree_df.shape)

question_df['statement'] = question_df['statement'].map(eval)
question_df['option_a'] = question_df['option_a'].map(eval)
question_df['option_b'] = question_df['option_b'].map(eval)
question_df['option_c'] = question_df['option_c'].map(eval)
question_df['option_d'] = question_df['option_d'].map(eval)

print(question_df.columns)
print(question_parse_tree_df.columns)
assert reference_df.shape[0] == reference_parse_tree_df.shape[0]

(26361, 9)
(26361, 7)
Index(['id', 'statement', 'option_a', 'option_b', 'option_c', 'option_d',
       'type', 'subject', 'answer'],
      dtype='object')
Index(['id', 'statement', 'option_a', 'option_b', 'option_c', 'option_d',
       'source'],
      dtype='object')


In [22]:
error_statement = []
error_option_a = []
error_option_b = []
error_option_c = []
error_option_d = []

question_parse_tree_dict = {question_parse_tree_df.loc[i, 'id']: (question_parse_tree_df.loc[i, 'statement'], 
                                                                  question_parse_tree_df.loc[i, 'option_a'],
                                                                  question_parse_tree_df.loc[i, 'option_b'], 
                                                                  question_parse_tree_df.loc[i, 'option_c'],
                                                                  question_parse_tree_df.loc[i, 'option_d']) for i in range(question_parse_tree_df.shape[0])}

for i in range(question_df.shape[0]):
    _id = question_df.loc[i, 'id']
    statement = question_df.loc[i, 'statement']
    option_a = question_df.loc[i, 'option_a']
    option_b = question_df.loc[i, 'option_b']
    option_c = question_df.loc[i, 'option_c']
    option_d = question_df.loc[i, 'option_d']
    
    statement_tree, option_a_tree, option_b_tree, option_c_tree, option_d_tree = question_parse_tree_dict[_id]   
    
    statement_from_parse_tree = generate_pos_tags_from_parse_tree(parse_tree=statement_tree)
    option_a_from_parse_tree = generate_pos_tags_from_parse_tree(parse_tree=option_a)
    option_b_from_parse_tree = generate_pos_tags_from_parse_tree(parse_tree=option_b)
    option_c_from_parse_tree = generate_pos_tags_from_parse_tree(parse_tree=option_c)
    option_d_from_parse_tree = generate_pos_tags_from_parse_tree(parse_tree=option_d)

    statement_ex = list(map(lambda x: x[1], statement_from_parse_tree))
    statement_ex = list(map(lambda x: '(' if x == '-LRB-' else ')' if x == '-RRB-' else '\xa0' if x == '' else x, statement_ex))
    if not content == content_ex:
        print('statement', i)
        error_statement.append(i)
        
    option_a_ex = list(map(lambda x: x[1], option_a_from_parse_tree))
    option_a_ex = list(map(lambda x: '(' if x == '-LRB-' else ')' if x == '-RRB-' else '\xa0' if x == '' else x, option_a_ex))
    if not content == content_ex:
        print('option_a', i)
        error_option_a.append(i)

    option_b_ex = list(map(lambda x: x[1], option_b_from_parse_tree))
    option_b_ex = list(map(lambda x: '(' if x == '-LRB-' else ')' if x == '-RRB-' else '\xa0' if x == '' else x, option_b_ex))
    if not content == content_ex:
        print('option_b', i)
        error_option_b.append(i)

    option_c_ex = list(map(lambda x: x[1], option_c_from_parse_tree))
    option_c_ex = list(map(lambda x: '(' if x == '-LRB-' else ')' if x == '-RRB-' else '\xa0' if x == '' else x, option_c_ex))
    if not content == content_ex:
        print('option_c', i)
        error_option_c.append(i)

    option_d_ex = list(map(lambda x: x[1], option_d_from_parse_tree))
    option_d_ex = list(map(lambda x: '(' if x == '-LRB-' else ')' if x == '-RRB-' else '\xa0' if x == '' else x, option_d_ex))
    if not content == content_ex:
        print('option_d', i)
        error_option_d.append(i)


print('题库中不能匹配的数量')
print(f'statement: {len(error_statement)}')
print(f'option_a: {len(error_option_a)}')
print(f'option_b: {len(error_option_b)}')
print(f'option_c: {len(error_option_c)}')
print(f'option_d: {len(error_option_d)}')

题库中不能匹配的数量
statement: 0
option_a: 0
option_b: 0
option_c: 0
option_d: 0


### 句法树生成networkx图测试

In [ ]:
def parse_trees_to_graph(parse_trees, succ_to, error_to):
    for parse_tree in parse_trees:
        try:
            graph = parse_tree_to_graph(parse_tree=parse_tree, display=False)
        except Exception as e:
            print(f'error: {parse_tree}')
            with open(error_to, 'a', encoding='utf8') as f:
                f.write(f'{parse_tree}\t{e}\n')

In [ ]:
df = pandas.read_csv(QUESTION_PARSE_TREE_PATH, sep='\t', header=0)

for i in range(df.shape[0]):
    option_a_tree = eval(df.loc[i, 'option_a'])
    option_b_tree = eval(df.loc[i, 'option_b'])
    option_c_tree = eval(df.loc[i, 'option_c'])
    option_d_tree = eval(df.loc[i, 'option_d'])
    
    parse_trees_to_graph(parse_trees=option_a_tree, succ_to='test_parse_tree_to_graph.txt', error_to='error_parse_tree_to_graph.txt')
    parse_trees_to_graph(parse_trees=option_b_tree, succ_to='test_parse_tree_to_graph.txt', error_to='error_parse_tree_to_graph.txt')
    parse_trees_to_graph(parse_trees=option_c_tree, succ_to='test_parse_tree_to_graph.txt', error_to='error_parse_tree_to_graph.txt')
    parse_trees_to_graph(parse_trees=option_d_tree, succ_to='test_parse_tree_to_graph.txt', error_to='error_parse_tree_to_graph.txt')

In [ ]:
df = pandas.read_csv(REFERENCE_PARSE_TREE_PATH, sep='\t', header=0)

for i in range(df.shape[0]):
    trees = eval(df.loc[i, 'content'])
    parse_trees_to_graph(parse_trees=trees, succ_to='test_parse_tree_to_graph.txt', error_to='error_parse_tree_to_graph.txt')

### 句法树分支数量情况（最大子女数量）

In [ ]:
# 确定最大子女数(参考文档)

df = pandas.read_csv(REFERENCE_PARSE_TREE_PATH, sep='\t', header=0)
logs = []
for i in range(df.shape[0]):
    content = eval(df.loc[i, 'content'])
    for tree_string in content:
        tree = parse_tree_to_graph(parse_tree=tree_string, display=False, return_type='networkx', ignore_text=False)
        edges = list(tree.edges)
        out_degrees = list(map(lambda x: x[0], edges))
        temp_max = Counter(out_degrees).most_common(1)[0][1]
        logs.append(temp_max)
print(max(logs))
print(sum(logs) / len(logs))

参考文档最大子女数为47，均值为3.245742220908256

In [ ]:
# 确定最大子女数(题库)

df = pandas.read_csv(QUESTION_PARSE_TREE_PATH, sep='\t', header=0)
max_children = 0

statement_logs = []
option_logs = []

for i in range(df.shape[0]):
    statement = eval(df.loc[i, 'statement'])
    option_a = eval(df.loc[i, 'option_a'])
    option_b = eval(df.loc[i, 'option_b'])
    option_c = eval(df.loc[i, 'option_c'])
    option_d = eval(df.loc[i, 'option_d'])
    options = option_a + option_b + option_c + option_d
    for tree_string in statement:
        tree = parse_tree_to_graph(parse_tree=tree_string, display=False, return_type='networkx', ignore_text=False)
        edges = list(tree.edges)
        out_degrees = list(map(lambda x: x[0], edges))
        temp_max = Counter(out_degrees).most_common(1)[0][1]
        statement_logs.append(temp_max)
    for tree_string in options:
        tree = parse_tree_to_graph(parse_tree=tree_string, display=False, return_type='networkx', ignore_text=False)
        edges = list(tree.edges)
        out_degrees = list(map(lambda x: x[0], edges))
        temp_max = Counter(out_degrees).most_common(1)[0][1]
        option_logs.append(temp_max)

print('statement:')
print(max(statement_logs))
print(sum(statement_logs) / len(statement_logs))
print('options:')
print(max(option_logs))
print(sum(option_logs) / len(option_logs))

题库中：
- 题干最大子女数为26，均值为3.0357229795520935
- 选项最大子女数为23，均值为2.8461790633467707

In [ ]:
tree.edges.to_list()

### 句法树在Dataset中的生成情况测试

1. 由于有的字符jieba可以识别而stanford句法解析不能识别，导致pos_tags的长度与分词编码长度不同，目前发现的特殊字符有：\xa0（编号为692）, \u3000（编号为26182）（在token2id.csv中是这个，但是这里的编码用的其实是reference_token2id.csv里的编码，那个token里面是没有\xa0和\u3000的），有点痛苦。
2. 20220518更新: \xa0可以被识别，但是单空格不能被stanford句法树识别，因此目前special_chars.txt中只有\u3000与单空格。
3. 20220518更新: 判定pos_tags有用字符数量的方式从`x!=''`改为`x in STANFORD_POS_TAG`, 更贴近原文

### use_reference为False的测试

In [ ]:
args = load_args(Config=DatasetConfig)
args.use_reference = False
args.retrieval_model_name = 'tfidf'
args.train_batch_size = 2
args.valid_batch_size = 2
args.test_batch_size = 2
args.use_parse_tree = True
args.use_pos_tags = True
args.word_embedding = None
args.document_embedding = None
pipeline = 'choice'
mode = 'test'

dataset = BasicDataset(args=args, mode=mode, do_export=False, pipeline=pipeline, for_debug=False)
df = dataset.data
print(df.columns)
print(df.shape)

In [ ]:
for i in range(df.shape[0]):
    options = df.loc[i, 'options']
    statement = df.loc[i, 'question']
    length_statement = len(list(filter(lambda x: x!=0 and x!='', statement)))
    length_a = len(list(filter(lambda x: x not in [0], options[0])))
    length_b = len(list(filter(lambda x: x not in [0], options[1])))
    length_c = len(list(filter(lambda x: x not in [0], options[2])))
    length_d = len(list(filter(lambda x: x not in [0], options[3])))
    
    statement_pos_tags = df.loc[i, 'statement_pos_tags']
    option_a_pos_tags = df.loc[i, 'option_a_pos_tags']
    option_b_pos_tags = df.loc[i, 'option_b_pos_tags']
    option_c_pos_tags = df.loc[i, 'option_c_pos_tags']
    option_d_pos_tags = df.loc[i, 'option_d_pos_tags']
    
#     length_statement_pos_tags = len(list(filter(lambda x: x!='', statement_pos_tags)))
#     length_a_pos_tags = len(list(filter(lambda x: x!='', option_a_pos_tags)))
#     length_b_pos_tags = len(list(filter(lambda x: x!='', option_b_pos_tags)))
#     length_c_pos_tags = len(list(filter(lambda x: x!='', option_c_pos_tags)))
#     length_d_pos_tags = len(list(filter(lambda x: x!='', option_d_pos_tags)))

    
    length_statement_pos_tags = len(list(filter(lambda x: x in STANFORD_POS_TAG, statement_pos_tags)))
    length_a_pos_tags = len(list(filter(lambda x: x in STANFORD_POS_TAG, option_a_pos_tags)))
    length_b_pos_tags = len(list(filter(lambda x: x in STANFORD_POS_TAG, option_b_pos_tags)))
    length_c_pos_tags = len(list(filter(lambda x: x in STANFORD_POS_TAG, option_c_pos_tags)))
    length_d_pos_tags = len(list(filter(lambda x: x in STANFORD_POS_TAG, option_d_pos_tags)))
    
    if not length_statement == length_statement_pos_tags:
        print(df.loc[i, 'id'], 'STATEMENT')
        print(statement, length_statement)
        print(statement_pos_tags, length_statement_pos_tags)
    
    if not length_a == length_a_pos_tags:
        print(df.loc[i, 'id'], 'A')
        print(options[0], length_a)
        print(option_a_pos_tags, length_a_pos_tags)
        
    if not length_b == length_b_pos_tags:
        print(df.loc[i, 'id'], 'B')
        print(options[1], length_b)
        print(option_b_pos_tags, length_b_pos_tags)
        
    if not length_c == length_c_pos_tags:
        print(df.loc[i, 'id'], 'C')
        print(options[2], length_c)
        print(option_c_pos_tags, length_c_pos_tags)

    if not length_d == length_d_pos_tags:
        print(df.loc[i, 'id'], 'D')
        print(options[3], length_d)
        print(option_d_pos_tags, length_d_pos_tags)

上面的测试无输出表明就正确

### use_reference为True的测试

In [ ]:
args = load_args(Config=DatasetConfig)
args.use_reference = True
args.retrieval_model_name = 'tfidf'
args.train_batch_size = 2
args.valid_batch_size = 2
args.test_batch_size = 2
args.use_parse_tree = True
args.use_pos_tags = True
args.word_embedding = None
args.document_embedding = None
pipeline = 'choice'
mode = 'train'

dataset = BasicDataset(args=args, mode=mode, do_export=False, pipeline=pipeline, for_debug=False)
df = dataset.data
print(df.columns)
print(df.shape)

In [ ]:
for i in range(df.shape[0]):
    reference = df.loc[i, 'reference']
    reference_pos_tags = df.loc[i, 'reference_pos_tags']
    length_reference = [len(list(filter(lambda x: x not in [0], _reference))) for _reference in reference]
    length_reference_pos_tags = [len(list(filter(lambda x: x in STANFORD_POS_TAG, _reference_pos_tags))) for _reference_pos_tags in reference_pos_tags]

    if not length_reference == length_reference_pos_tags:
        print(df.loc[i, 'id'], 'REFERENCE')
        print(length_reference, length_reference_pos_tags)
    

上面的测试无输出表明就正确

## 依存关系图

In [ ]:
reference_dependency_df = pandas.read_csv(REFERENCE_DEPENDENCY_PATH, sep='\t')

In [ ]:
reference_dependency_df

In [ ]:
count = 0

errors = []

for dependency in reference_dependency_df['content']:
    if not str(dependency)[0] == '[':
        errors.append(str(dependency))
        count += 1

print(count)
print(Counter(errors))

错误分析：
- 'NoneType' object is not subscriptable: .triples()方法出错，这个问题很棘手，目前理解为解析错误，但是不知道具体怎么解决。
- Number of tab-delimited fields (9) not supported by CoNLL(10) or Malt-Tab(4) format: 由于\xa0特殊字符，需要删除
- nan: 发生了断言错误，据观察也是与\xa0，这个出现问题的都是句中连续的\xa0，前一个则是句首。

In [ ]:
reference_dependency_df[reference_dependency_df['content'].isna()]

# BERT输出本地化的一些测试

已经确认的事实：

- reference_bert_output共计24718行，全部都成功生成了BERT张量
- question_bert_output共计26361行，全部都成功生成了BERT张量
- 题库中训练集和测试集上的题目id没有重复的情况

In [ ]:
reference_bert_outputs = dill.load(open(REFERENCE_BERT_OUTPUT_PATH, 'rb'))
question_bert_outputs = dill.load(open(QUESTION_BERT_OUTPUT_PATH, 'rb'))

In [ ]:
len(set(question_bert_outputs['id']))

In [ ]:
len(question_bert_outputs['id'])

# 图模型测试

In [ ]:
reference_parse_tree_df = pandas.read_csv(REFERENCE_PARSE_TREE_PATH, sep='\t')